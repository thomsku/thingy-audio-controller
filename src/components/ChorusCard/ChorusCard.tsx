import React from "react";

interface IProps {
  buttonPressed: boolean;
}

const ConnectButton: React.FunctionComponent<IProps> = (props) => {
  return (
    <div className="card">
      <p>press the button</p>
      {props.buttonPressed ? (
        <p>chorus is <span className="on">on</span></p>
      ) : (
        <p>chorus is <span className="off">off</span></p>
      )}
    </div>
  );
};

export default ConnectButton;
