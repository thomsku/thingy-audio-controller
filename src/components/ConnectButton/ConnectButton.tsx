import React from "react";

interface IProps {
  connected: boolean;
  connect(): void;
}

const ConnectButton: React.FunctionComponent<IProps> = (props) => {
  return (
    <div className="buttondiv">
      <button className="connectbutton" onClick={props.connect}>
        {props.connected ? (
          "disconnect"
        ) : (
          "connect"
        )}
      </button>
    </div>
  );
};

export default ConnectButton;
