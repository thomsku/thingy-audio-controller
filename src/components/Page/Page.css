.page {
  padding-top: 1em;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

.page > .buttondiv {
  height: 3em;
}

