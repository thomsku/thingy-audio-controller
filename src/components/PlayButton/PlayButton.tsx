import React from "react";

interface IProps {
  playingAudio: boolean;
  play(): void;
}

const ConnectButton: React.FunctionComponent<IProps> = (props) => {
  return (
    <div className="buttondiv">
      <button className="playbutton" onClick={props.play}>
        {props.playingAudio ? (
          "stop"
        ) : (
          "play"
        )}
      </button>
    </div>
  );
};

export default ConnectButton;
