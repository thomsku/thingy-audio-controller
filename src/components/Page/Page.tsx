import React from "react";
import Thingy from "thingy52_web_bluetooth";
import Tone from "tone";
import ConnectButton from "../ConnectButton/ConnectButton";
import PlayButton from "../PlayButton/PlayButton";
import "./Page.css";

interface IState {
  connected: boolean;
  playingAudio: boolean;
  temperature: number;
}

class Page extends React.Component<{}, IState> {
  public state: IState = {
    connected: false,
    playingAudio: false,
    temperature: -1,
  };
  private player!: Tone.Player;
  private filter!: Tone.Filter;
  private chorus!: Tone.Chorus;

  public componentDidMount() {
    this.chorus = new Tone.Chorus(2, 2.5, 0.0).toMaster();
    // this.chorus.wet.value = 0;
    this.filter = new Tone.Filter(1500, "lowpass").toMaster();
    this.player = new Tone.Player("./song.mp3");
    this.player.loop = true;
    this.player.connect(this.chorus);
  }

  public connect = async () => {
    if (!this.state.connected) {
      try {
        const thingy = new Thingy({logEnabled: false});
        (window as any).thingy = thingy;
        const connected = await (window as any).thingy.connect();
        if (connected) {
          (window as any).thingy.addEventListener("button", this.temperatureToFilterListener);
          await (window as any).thingy.button.start();
          this.setState({
            connected: true,
          });
        }
      } catch (e) {
        throw(e);
      }
    } else {
      await (window as any).thingy.disconnect();
      this.setState({
        connected: false,
      });
    }
  }

  public play = () => {
    if (this.state.playingAudio === false) {
      this.player.start();
    } else {
      this.player.stop();
    }
    this.setState({
      playingAudio: (!this.state.playingAudio),
    });
  }

  public temperatureToFilterListener = (data: any) => {
    // can do gravity with filter?
    if (data.detail.value === 1) {
      this.chorus.depth = 0.5;
    } else {
      this.chorus.depth= 0;
    }
    // this.filter.frequency.value = data.detail.heading * 30 + 300;
  }

  public render() {
    return (
    <div className="page">
      <h1>thingy audio controller</h1>
      <ConnectButton connect={this.connect} connected={this.state.connected}/>
      <PlayButton play={this.play} playingAudio={this.state.playingAudio}/>
    </div>
  );
  }
}

export default Page;
