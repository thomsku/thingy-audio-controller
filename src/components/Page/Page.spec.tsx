import { mount, ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import React from "react";
import ChorusCard from "../ChorusCard/ChorusCard";
import ConnectButton from "../ConnectButton/ConnectButton";
import PlayButton from "../PlayButton/PlayButton";
import Page from "./Page";

let mockWebBluetooth = true;

jest.mock("thingy52_web_bluetooth", () => {
// tslint:disable-next-line: only-arrow-functions
  return function() {
    return {
      addEventListener: jest.fn(),
      button: {start: jest.fn()},
      connect: () => (mockWebBluetooth ? true : false),
      disconnect: () => true,
      temperature: {start: jest.fn()},
    };
  };
});

jest.mock("tone", () => {
  return {
    Chorus: jest.fn().mockImplementation((frequency, delayTime, depth) => {
      return {
        delayTime,
        depth,
        frequency,
        toMaster: jest.fn(),
        wet: {value: 1},
      };
    }),
    Filter: jest.fn().mockImplementation((frequency, type) => {
      return {
        frequency,
        toMaster: jest.fn(),
        type,
      };
    }),
    Player: jest.fn().mockImplementation(() => {
      return {
        connect: jest.fn(),
        start: jest.fn(),
        stop: jest.fn(),
        toMaster: jest.fn(),
      };
    }),
  };
});

// jest.genMockFromModule("tone");
// jest.mock("tone");

describe("Page", () => {
  let wrapper: ShallowWrapper<null, null, Page>;
  beforeEach(() => wrapper = shallow(<Page />));

  it("should render correctly", () => expect(wrapper).toMatchSnapshot());

  it("should render a <div />", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });

  it("should render the ConnectButton and PlayButton Components", () => {
    expect(wrapper.containsAllMatchingElements([
    <ConnectButton
      connect={wrapper.instance().connect}
      connected={wrapper.instance().state.connected}/>,
    <PlayButton
      play={wrapper.instance().play}
      playingAudio={wrapper.instance().state.playingAudio}/>,
    <ChorusCard
      buttonPressed={wrapper.instance().state.buttonPressed}/>,
    ])).toEqual(true);
  });
});

describe("mounted Page", () => {
  let wrapper: ReactWrapper<null, null, Page>;
  beforeEach(() => wrapper = mount(<Page />));

  it("should call connect when the connect button is clicked", () => {
    const spy = jest.spyOn(wrapper.instance(), "connect");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find(".connectbutton").simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should call play when the play button is clicked", () => {
    const spy = jest.spyOn(wrapper.instance(), "play");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find(".playbutton").simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });
});

describe("connect", () => {
  let wrapper: ReactWrapper<null, null, Page>;
  beforeEach(() => wrapper = mount(<Page />));

  it("should update the text of the ConnectButton Component based on connected state", async () => {
    wrapper.setState({connected: false});
    await wrapper.instance().connect();
    await wrapper.instance().forceUpdate();
    expect(wrapper.instance().state.connected).toEqual(true);
    expect(wrapper.find(".connectbutton").text()).toEqual("disconnect");
    await wrapper.instance().connect();
    await wrapper.instance().forceUpdate();
    expect(wrapper.instance().state.connected).toEqual(false);
    expect(wrapper.find(".connectbutton").text()).toEqual("connect");
  });

  it("should start the button listener on Thingy", async () => {
    const spy = jest.spyOn((window as any).thingy.button, "start");
    wrapper.setState({connected: false});
    await wrapper.instance().connect();
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should not connect when the browser does not support Web Bluetooth", async () => {
    mockWebBluetooth = false;
    wrapper.setState({connected: false});
    await wrapper.instance().connect();
    wrapper.instance().forceUpdate();
    expect(wrapper.instance().state.connected).toEqual(false);
  });

});

describe("play", () => {
  let wrapper: ReactWrapper<null, null, Page>;
  beforeEach(() => wrapper = mount(<Page />));

  it("should update the text of the PlayButton Component based on playing state", () => {
    wrapper.setState({playingAudio: false});
    wrapper.instance().play();
    wrapper.instance().forceUpdate();
    expect(wrapper.instance().state.playingAudio).toEqual(true);
    expect(wrapper.find(".playbutton").text()).toEqual("stop");
    wrapper.instance().play();
    wrapper.instance().forceUpdate();
    expect(wrapper.instance().state.playingAudio).toEqual(false);
    expect(wrapper.find(".playbutton").text()).toEqual("play");
  });
});

describe("toggleChorus", () => {
  let wrapper: ReactWrapper<null, null, Page>;
  beforeEach(() => wrapper = mount(<Page />));

  it("should change the depth of the chorus based on the button state", () => {
    wrapper.instance().componentDidMount();
    wrapper.instance().toggleChorus({detail: {value: 1}});
    wrapper.instance().forceUpdate();
    expect(wrapper.instance().getChorus().wet.value).toEqual(1);
    wrapper.instance().toggleChorus({detail: {value: 0}});
    wrapper.instance().forceUpdate();
    expect(wrapper.instance().getChorus().wet.value).toEqual(0);
  });
});
