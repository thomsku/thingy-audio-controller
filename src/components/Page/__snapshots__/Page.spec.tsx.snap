// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page should render correctly 1`] = `ShallowWrapper {}`;
